server/src/main/java/org/elasticsearch/index/shard/IndexWriteLoadForecast.java

```diff
@@ -0,0 +1,162 @@
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.OptionalDouble;
+import java.util.OptionalLong;
+
+public class IndexWriteLoadForecast implements Writeable, ToXContentFragment {
+    public static final ParseField SHARDS_WRITE_LOAD_FIELD = new ParseField("loads");
+
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<IndexWriteLoadForecast, Void> PARSER = new ConstructingObjectParser<>(
+        "index_write_load_forecast_parser",
+        false,
+        (args, unused) -> new IndexWriteLoadForecast(((List<Double>) args[0]).stream().mapToDouble(value -> value).toArray())
+    );
+
+    static {
+        PARSER.declareDoubleArray(ConstructingObjectParser.constructorArg(), SHARDS_WRITE_LOAD_FIELD);
+    }
+
+    private final double[] forecastedShardWriteLoad;
+
+    private IndexWriteLoadForecast(double[] forecastedShardWriteLoad) {
+        this.forecastedShardWriteLoad = forecastedShardWriteLoad;
+    }
+
+    public IndexWriteLoadForecast(StreamInput in) throws IOException {
+        this(in.readDoubleArray());
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeDoubleArray(forecastedShardWriteLoad);
+    }
+
+    public static ClusterState maybeIncludeWriteLoadForecast(
+        IndexAbstraction.DataStream dataStream,
+        ClusterState clusterState,
+        TimeValue maxIndexAge,
+        TimeValue minShardUptime
+    ) {
+        final IndexMetadata writeIndex = clusterState.metadata().getIndexSafe(dataStream.getWriteIndex());
+        if (IndexSettings.FORECAST_WRITE_LOAD_SETTING.get(writeIndex.getSettings()) == false) {
+            return clusterState;
+        }
+
+        final int numberOfShards = writeIndex.getNumberOfShards();
+        final double[] totalWriteLoad = new double[numberOfShards];
+        final double[] maxWriteLoad = new double[numberOfShards];
+        final int[] numberOfIndicesTookIntoAccount = new int[numberOfShards];
+
+        for (Index index : dataStream.getIndices()) {
+            final IndexMetadata indexMetadata = clusterState.metadata().getIndexSafe(index);
+            final long indexAge = System.currentTimeMillis() - indexMetadata.getCreationDate();
+            if (index.equals(dataStream.getWriteIndex()) || indexAge > maxIndexAge.millis()) {
+                continue;
+            }
+
+            final IndexWriteLoad writeLoad = indexMetadata.getWriteLoad();
+            for (int shardId = 0; shardId < numberOfShards; shardId++) {
+                OptionalDouble writeLoadForShard = writeLoad.getWriteLoadForShard(shardId);
```

Comment:
This can fail if the number of shards differ. Also, we may not consider all shards if the number of shards is now less than before.